Add some challenges

wielandb opened the pull request:
This PR adds one proper challenge and three drafts for ones that could exist in the future.

wielandb committed:
Initial commit

wielandb committed:
Change names to fit intended structure

wielandb committed:
Add READMEs

wielandb committed:
Change challenge description

wielandb committed:
Remove comment

Robbendebiene commented on challenges/stop_area-names-from-platform-names/relationcheck.py:
```diff
@@ -0,0 +1,67 @@
+platform_names = set()
+
+def fetch_relation_data(relation_id):
+    result = requests.get(f"https://overpass-api.de/api/interpreter?data=relation%28" + str(relation_id) + "%29%3B%28._%3B%3E%3B%29%3Bout%3B")
```
For all your code: Use functions to build urls which do the encoding of the parameters (pretty sure python has tools for this) so the url in code is human readable.

wielandb replied:
Should be resolved

Robbendebiene reviewed (changes requested):
Requested changes:
- no preset-README. Do a proper README or remove the directory
- no presets directory in general. We are programmers after all, we want to reuse code and not duplicate it. This is hell to maintain. Instead re-use code by putting it into little components such as functions, classes etc. (this is what the `shared` dir is for)
- create reusable classes like "Feature" and "Task" that implement a "toJSON" methods where the json structure is constructed.  Also move the overpass query calls like "get", toJson etc. into a shared function that you can use in other challenges
- do not upload the task.json to github

7h30n3 commented on challenges/stop_area-names-from-platform-names/relationcheck.py:
```diff
@@ -0,0 +1,67 @@
+        f.write(data)
+    result = check_platform_names(relation_id, "data.osm")
+    if result is None:
+        print("Ergebnis: Keine Platform hatte einen Namen, wir können nichts tun")
```
In Germany we call it "Plattform".

wielandb committed:
Rework stop_area-name challenge

wielandb committed:
Fix a lat/lon switcheroony

wielandb committed:
Remove switch that ignores relations

wielandb committed:
Refactor code

wielandb committed:
Refactor Maproulette Challenge Factory

Robbendebiene commented on challenges/large-stop_area-bbox/main.py:
```diff
@@ -0,0 +1,67 @@
+    dlon = lon2 - lon1
+    a = math.sin(dlat / 2) * math.sin(dlat / 2) + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) * math.sin(dlon / 2)
+    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
+    return R * c
```
Please import a library that does geodesic calculations like https://pypi.org/project/geopy/

wielandb replied:
Done

Robbendebiene commented on challenges/large-stop_area-bbox/main.py:
```diff
@@ -0,0 +1,67 @@
+		t = mrf.Task(
+			mainFeature=mainFeature, 
+			additionalFeatures=[], 
+			cooperativeWork=None)
```
Make them optional parameters by assigning the empty list and None by default inside the Task constructor so you do not have to pass them when they are not needed. Not sure if the same can be done for `GeoFeature`

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+			print(geometry)
+			raise ValueError("Invalid coordinates")
+
+	def to_dict(self):
```
please name these functions "toJSON" or "toGeoJSON" or something like this for all the classes and convert to camel case.

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+	# The geometry is a dict with a type and coordinates
+	# The properties is a dict with key-value pairs
+	def __init__(self, geometry, properties = {}):
+		self.geometry = geometry
```
The geometry you pass here should already contain the "type", making all these if checks below obsolete.

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+
+	def to_dict(self):
+		# the features are the main feature and the additional features as one list
+		features = [self.mainFeature.to_dict()]
```
I think you can do
```
features = [
   self.mainFeature.to_dict()
   *map(lambda f: f.to_dict(), self.additionalFeatures)
 ]
```

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+		features = [self.mainFeature.to_dict()]
+		for feature in self.additionalFeatures:
+			features.append(feature.to_dict())
+		if self.cooperativeWork == None:
+			return {
+				"type": "FeatureCollection",
+				"features": features,
+			}
+		else:
+			return {
+				"type": "FeatureCollection",
+				"features": features,
+				"cooperativeWork": self.cooperativeWork.to_dict()
+			}
```
I'm not sure if python supports something like this:

```
		return {
				"type": "FeatureCollection",
				"features": features,
				if (self.cooperativeWork) "cooperativeWork": self.cooperativeWork.to_dict()
			}
```

Alternateivly you can use:

		
```
			dict = {
				"type": "FeatureCollection",
				"features": features,
			}
                       if self.cooperativeWork:
                          dict["cooperativeWork"] = self.cooperativeWork.to_dict()
                       return dict;
```

wielandb replied:
I did something similar

wielandb replied:
```
        return {
            "type": "FeatureCollection",
            "features": features,
            **({"cooperativeWork": self.cooperativeWork.toGeoJSON()} if self.cooperativeWork != None else {})
        }
```

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+class Overpass:
+	def __init__(self):
+		self.resultElements = {}
+		self.overpass_url = "https://overpass-api.de/api/interpreter"
```
would be nice if this could be **optionally** set from the outside in the constructor

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+
+class Overpass:
+	def __init__(self):
+		self.resultElements = {}
```
obsolete

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+		self.resultElements = {}
+		self.overpass_url = "https://overpass-api.de/api/interpreter"
+
+	def getElementsFromQuery(self, overpass_query):
```
Nit: perhaps "queryElements(...)" describes the function as well

Robbendebiene commented on include/MRFactory.py:
```diff
@@ -0,0 +1,215 @@
+		return resultElements
+
+
+def createElementCenterPoint(element):
```
I still believe it is worth thinking about a "Geometry" class. The idea is to create the true geometry and then allow to get simplifications of it via functions like "center" oder "bbox". Again for the bbox and center methods I recommend using a geo library.

```
class Geometry {
   constructor fromElement();
   center();
   bbox();
}
```

Robbendebiene reviewed (changes requested):
- Rename the "include" directory to "shared"
- Again please remove the generated json files.
- Use snake case for directories and files + only english

Otherwise the class code is nicely clean. Good job!

wielandb committed:
Remove MainGeoFeature

wielandb committed:
Factory -> ChallengeBuilder

wielandb committed:
Add Convert option to GeoFeature

wielandb committed:
Create ZHV functionality

wielandb committed:
Do not filter away train stations in large_area_bb

wielandb committed:
Change filter distance

wielandb committed:
Style the ZHV marker

wielandb committed:
Convert tabs to spaces

wielandb committed:
Move and remove files

wielandb committed:
Change lib file name

wielandb committed:
Add support for self-made geometry

wielandb committed:
Make large_area_bboxes consider all members of the stop_area relation

Even members that may be relations themselves

wielandb committed:
Fix polygon implementation

wielandb committed:
include -> shared

wielandb commented:
> * Rename the "include" directory to "shared"
Done
> * Again please remove the generated json files.
Added to .gitignore and deleted
> * Use snake case for directories and files + only english
Done

wielandb committed:
Use a lib for geographic calculations

wielandb committed:
to_dict -> toGeoJSON

wielandb committed:
Use lambda calculus for list generation

wielandb committed:
Use dictionary comprehension

wielandb committed:
Make the overpass url changable

wielandb committed:
Remove obsolete line

wielandb committed:
Update README.md

wielandb committed:
First Workflow

wielandb committed:
Set PYTHONPATH in actions

wielandb committed:
Reformat action

wielandb committed:
Fix attempt 2

wielandb committed:
Reset python path

wielandb committed:
Fix attempt 3

wielandb committed:
import requests

wielandb committed:
Be less verbose

wielandb committed:
Fix import for other challenge

wielandb committed:
Take the JSONs from root

wielandb committed:
Fix #4

wielandb committed:
Reintroduce JSON

wielandb committed:
Make the JSON files single artifacts

wielandb committed:
First Geo attempt

wielandb committed:
Fix paths, implement Geometry into one challenge

wielandb committed:
Revert "Fix paths, implement Geometry into one challenge"

This reverts commit 3c53817f80cd5c852e810d0e9b8d8cc578ce1d7d.

wielandb committed:
Make the tag name static

wielandb committed:
Make the path static from everywhere

wielandb committed:
Make the release name and tag the date

wielandb committed:
Update README.md

wielandb committed:
fix spelling errors

wielandb committed:
Make the scripts use the new geometry class

wielandb commented:
Should be ready for review now

wielandb committed:
Fix a typo

Robbendebiene commented on README.md:
```diff
@@ -1,2 +1,18 @@
+For usage with MapRoulette, it is recommended that you use the static url that will always point to the latest release.
+These URLs are currently:
+- https://github.com/wielandb/PT-QA-MapRoulette/releases/latest/download/large_stop_area_bbox.json
+- https://github.com/wielandb/PT-QA-MapRoulette/releases/latest/download/stop_area_names_from_platform_names.json
```
Please adjust the urls accordingly. Currently they are all pointing to your fork.

Robbendebiene commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,74 @@
+        upload_url: ${{ steps.create_release.outputs.upload_url }}
+        asset_path: artifacts/stop_area_names_from_platform_names.json
+        asset_name: stop_area_names_from_platform_names.json
+        asset_content_type: application/json
```
Please use [this](https://github.com/softprops/action-gh-release) action as the others are unmaintained.

This should also allow you to simplify the script.

I would like the script to automatically upload all files inside the `artifacts` directory (without exact specification).

wielandb replied:
Seems to have worked!

Robbendebiene commented on challenges/large_stop_area_bbox/main.py:
```diff
@@ -0,0 +1,67 @@
+from pathlib import Path
+sys.path.append(str(Path(__file__).parent.parent.parent / "shared"))
+#from ..shared import challenge_builder as mrcb
+#import challenge_builder as mrcb
```
Is this obsolete code?

Robbendebiene commented on challenges/large_stop_area_bbox/main.py:
```diff
@@ -0,0 +1,67 @@
+foreach {
+  >> -> .ancestors;
+  make myCustomElement
+    ::id=min(id()),
+  	::geom=ancestors.gcat(geom());
```
Minor: formatting

wielandb replied:
Huh... it doesn't look like this for me... :o

wielandb replied:
Solved... my tab-distance was off

Robbendebiene replied:
This usually happens when using tabs, please ensure you are only using spaces.

Robbendebiene commented on challenges/stop_area_names_from_platform_names/README.md:
```diff
@@ -0,0 +1,20 @@
+# Add names to stop_areas by platform names
+
+This script gets all stop_areas without names from the overpass API. It then fetches every relation from the OSM API and checks if all platforms in this relation have the same name. If they do, this name is suggested to be added as the name for the whole stop_area.
```
The query behind has changed a little bit. Please adjust the description accordingly.

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -0,0 +1,299 @@
+            for point in self.coordinates[0]:
+                lat += point[1]
+                lon += point[0]
+            return [lon / len(self.coordinates[0]), lat / len(self.coordinates[0])]
```
I would like to do more work with a geodesic python package. My suggestion is to swap geopy with https://turfpy.readthedocs.io/en/latest/turfpy.html This allows distance calculation, while also allowing you to calculate the centroid or bbox for geojson objects. (see: https://turfpy.readthedocs.io/en/latest/turfpy.measurement.html)

This also has the benefit that we can do a lot more complicated calculation (like intersections etc.) in the future.

Ideally you work with the geojson package (https://pypi.org/project/geojson/) as well. The only thing you really need is the `fromOverpassElement`

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -0,0 +1,299 @@
+    def __init__(self, geometry, properties = {}):
+        self.geometry = geometry
+        self.properties = properties
+        # infer the type of the geometry (Point, LineString, Polygon) from the coordinates
+        # if its just a list of 2 numbers, its a Point
+        # if its a list of lists of 2 numbers, its a LineString
+        # if its a list of lists of 2 numbers where start and end are the same, its a Polygon
```
This comment is no longer valid. At least not at this place.

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -0,0 +1,299 @@
+
+    @classmethod
+    def withId(cls, osmType, osmId, geometry, properties):
+        print(osmType, osmId, geometry, properties)
```
Please remove any prints from the shared library

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -0,0 +1,299 @@
+            "properties": self.properties
+        }
+
+    def convertPolygonToClosedString(self):
+        if self.geometryType != "Polygon":
+            raise ValueError("This function only works for Polygons")
+        self.geometryType = "LineString"
+        self.geometry = self.geometry[0]
+
+    def convertClosedStringToPolygon(self):
+        if self.geometryType != "LineString":
+            raise ValueError("This function only works for LineStrings")
+        self.geometryType = "Polygon"
+        self.geometry = [self.geometry]
```
Please remove these functions.
Geometry manipulations should directly happen on the geometry object.

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -0,0 +1,299 @@
+                lon += point[0]
+            element['lat'] = lat / len(element['geometry']['coordinates'])
+            element['lon'] = lon / len(element['geometry']['coordinates'])
+    else:
+        print(element)
+        raise ValueError("No handalable coordinates found for element")
+    return element
+
+def createGeometryFromElement(element):
+    # "geometry" needs to check before "bounds", because when you output geom, bbox will also be given back
+    # if there is "bounds" in element, construct a polygon with 4 points that represents the bbox
+    if "lat" in element:
+        # list of 2 floats for a dict that has lat lon as keys
+        element["simpleGeometry"] = [element["lon"], element["lat"]]
+    elif "geometry" in element and "coordinates" in element["geometry"]:
+        # list of lists of 2 floats for a dict that has lat lon as keys
+        element["simpleGeometry"] = element["geometry"]["coordinates"]
+    elif "geometry" in element:
+        # list of lists of 2 floats for a dict that has lat lon as keys
+        element["simpleGeometry"] = [[point["lon"], point["lat"]] for point in element["geometry"]]
+    elif "bounds" in element:
+        element["simpleGeometry"] = [
+            [element["bounds"]["minlon"], element["bounds"]["minlat"]],
+            [element["bounds"]["minlon"], element["bounds"]["maxlat"]],
+            [element["bounds"]["maxlon"], element["bounds"]["maxlat"]],
+            [element["bounds"]["maxlon"], element["bounds"]["minlat"]],
+            [element["bounds"]["minlon"], element["bounds"]["minlat"]]
+        ]	
+    else:
+        print(element)
+        raise ValueError("No handalable geometry found for element")
+    return element
+
+def getElementCenterPoint(element):
+    newElement = createElementCenterPoint(element)
+    return [newElement["lon"], newElement["lat"]]
+
+def getElementGeometry(element):
+    newElement = createGeometryFromElement(element)
+    return newElement["simpleGeometry"]
```
Should all be obsolete.
The working principle should always be:
1. convert Overpass results to geometry / geojson instance
2. Manipulate the geometry/Create a geometry if needed via turf package functions
3. Pass the desired geometry to the GeoFeature

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -0,0 +1,299 @@
+    def __init__(self, overpass_url = "https://overpass-api.de/api/interpreter"):
+        self.overpass_url = overpass_url
+
+    def getElementsFromQuery(self, overpass_query):
```
name to `queryElements()`

Robbendebiene commented on challenges/large_stop_area_bbox/README.md:
````diff
@@ -0,0 +1,20 @@
+
+
+```
+
````
Please update this readme.

wielandb committed:
Update README with new URLs

wielandb committed:
Merge branch 'main' of https://github.com/wielandb/PT-QA-MapRoulette

wielandb committed:
Update GH Action

wielandb committed:
Update release.yml

wielandb committed:
Reformat indentations

wielandb committed:
Update README.md

wielandb committed:
Remove other imports

wielandb committed:
Update README.md

wielandb committed:
Rework challenge builder

wielandb committed:
getElementsFromQuery -> queryElements

Robbendebiene commented on README.md:
```diff
@@ -8,11 +8,11 @@ Public transport quality assurance callenges for MapRoulette
-These URLs are currently:
-- https://github.com/wielandb/PT-QA-MapRoulette/releases/latest/download/large_stop_area_bbox.json
-- https://github.com/wielandb/PT-QA-MapRoulette/releases/latest/download/stop_area_names_from_platform_names.json
+Theese URLs are currently:
```
It was correct before 🙈

```suggestion
These URLs are currently:
```

Robbendebiene commented on challenges/stop_area_names_from_platform_names/README.md:
```diff
@@ -1,6 +1,6 @@
 # Add names to stop_areas by platform names
 
-This script gets all stop_areas without names from the overpass API. It then fetches every relation from the OSM API and checks if all platforms in this relation have the same name. If they do, this name is suggested to be added as the name for the whole stop_area.
+This script gets all stop_areas without names from the overpass API. In the query, a virtual OSM object is constructed with a "combined name" of all platforms of this stop_area. If they are different, the `name=`-tag of this new object will read `< multiple values found >`, and it will not be considered. If they are all the name, this name is written to the `name=`-tag and suggested to be added a ...
```
Typo + minor adjustments:

```suggestion
This script gets all stop_areas without names from the overpass API. In the query, a virtual OSM object is constructed with a "combined name" of all platforms of this stop_area. If their names are different, the `name=`-tag of this new object will read `< multiple values found >`, and it will not be considered. If their names are equal, this name is written to the `name=`-tag and suggested to be added as the name for the whole stop_area.
```

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -2,77 +2,55 @@
-            if 'lat' in element:
-                GeomType = "Point"
-            elif 'center' in element:
+    def fromOverpassElement(cls, element, GeomType=None):
```
minor: do not start variable names with an upper case letter. Use `geomType`

Robbendebiene commented on challenges/large_stop_area_bbox/main.py:
```diff
@@ -1,8 +1,6 @@
-#from ..shared import challenge_builder as mrcb
-#import challenge_builder as mrcb
 import challenge_builder as mrcb
 from geopy import distance
```
geopy should be replace by turf completely. Also remove it from the requirements.

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -2,77 +2,55 @@
+from turfpy.measurement import distance, bbox, centroid
 
 @dataclass
 class Geometry:
```
I very much tried to avoid it, but it just makes things more simple so...

Please only keep the code of the `fromOverpassElement(element, GeomType=None):` function and make it return direct [geojson objects](https://pypi.org/project/geojson/#geojson-objects).
Example:
```py
        if GeomType == "Point":
            if 'geometry' in element:
                return geojson.Point(element['geometry']['coordinates'])
```
Remove the entire Geometry class with all it other functions.


`GeoFeature` can then extend the `Feature` class from https://pypi.org/project/geojson/ wherefore the only thing you have to implement is

```
    @classmethod
    def withId(cls, osmType, osmId, geometry, properties):
        properties["@id"] = f"{osmType}/{osmId}"
```
because the rest should already be done by the `Feature` class.


Finally we adjust the methods of the `Overpass` class

- rename: `queryElements` ->` queryElementsRaw`
- add: `queryElementsAsGeoJSON()`  - this internally calls `queryElementsRaw` and then contains all the code from `Geometry.fromOverpassElement` so that it returns a list of geojson objects

_____


If you want to get the centroid in any of your challenge scripts you can call `centroid(geojson_obj)` 
The functions `convertPolygonToClosedString` and `convertClosedStringToPolygon` are not as readable anymore but still one-liners:
-  `convertPolygonToClosedString`  -> `geojson.LineString(polygon.coordinates[0])`
- `convertClosedStringToPolygon` ->`geojson.Polygon([lineString.coordinates])`

wielandb committed:
Update README.md

Co-authored-by: Robin Thomas <Robbendebiene@users.noreply.github.com>

wielandb committed:
Update challenges/stop_area_names_from_platform_names/README.md

Co-authored-by: Robin Thomas <Robbendebiene@users.noreply.github.com>

wielandb committed:
Remove Geometry Class in favour of geojson.Geometry

wielandb committed:
GeomType -> forceGeomType / geopy -> turfpy

wielandb committed:
Update requirements.txt

wielandb committed:
Fix one challenge not beeing generated correctly

wielandb committed:
Refactor a function into Overpass Class

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -56,6 +18,9 @@ def __init__(self, geometry, properties={}):
     @classmethod
     def withId(cls, osmType, osmId, geometry, properties):
         properties["@id"] = f"{osmType}/{osmId}"
+        # Fix polygons beeing not a 3-dimensional array
+        if geometry["type"] == "Polygon" and len(geometry["coordinates"]) != 1:
+            geometry["coordinates"] = [geometry["coordinates"]]
```
This shouldn't be part of this class. Please move this to another place (e.g. queryElementsAsGeoJSON)

wielandb replied:
Moved it to init of the same class. Everywhere else, it has no effect.

Robbendebiene commented on challenges/stop_area_names_from_platform_names/main.py:
```diff
@@ -28,21 +28,21 @@
 
 for element in resultElements:
-    centerPoint = mrcb.geoJSONGeometryFromOverpassElement(element, forceGeomType="Point")
+    #centerPoint = element
```
Remove code that is commented out

wielandb committed:
Make queryElementsAsGeoJSON do all the Feature work

wielandb committed:
Remove print statements

wielandb committed:
move geometry fixing to somewhere else

wielandb committed:
Move a comment

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -90,7 +90,9 @@ def queryElementsRaw(self, overpass_query):
-        return [self.geoJSONGeometryFromOverpassElement(element, forceGeomType) for element in self.queryElementsRaw(overpass_query)]
+        rawElements = self.queryElementsRaw(overpass_query)
+        geometriesAndProperties = [(self.geoJSONGeometryFromOverpassElement(element, forceGeomType)) for element in rawElements]
+        return [geojson.Feature(geometry=g, properties=p) for g,p in geometriesAndProperties]
```
`geoJSONGeometryFromOverpassElement` I don't get why a function named `geoJSONGeometryFromOverpassElement` should return a tuple containing the tags as well.

Read the tags inside `queryElementsAsGeoJSON` and combine these two for loops into one.

Robbendebiene replied:
I see that you don't want to move the code there, then please split up the function into two. Functions should have a single responsible if feasible.

Create the following function and use it inside `queryElementsAsGeoJSON`
```python
geoJSONPropertiesFromOverpassElement()
    tags = element["tags"] or {}
    tags["@type"] = element["type"]
    tags["@id"] = element["id"]
    return tags;
```

wielandb replied:
Done like that

Robbendebiene commented on shared/challenge_builder.py:
```diff
@@ -92,7 +98,8 @@ def queryElementsRaw(self, overpass_query):
         rawElements = self.queryElementsRaw(overpass_query)
         geometriesAndProperties = [(self.geoJSONGeometryFromOverpassElement(element, forceGeomType)) for element in rawElements]
-        return [geojson.Feature(geometry=g, properties=p) for g,p in geometriesAndProperties]
+        featureList = [geojson.Feature(geometry=g, properties=p) for g,p in geometriesAndProperties]
+        return featureList
```
Why adding another variable instead of the direct return as before?

Robbendebiene replied:
If you want to go this way use a [map lambda function](https://www.geeksforgeeks.org/python-map-function/).

wielandb replied:
Should be good now

wielandb committed:
Only have one for loop in queryElementsAsGeoJSON

wielandb committed:
Rework some functions

wielandb committed:
return things directly
